os_traits/const.py

```diff
@@ -67,9 +67,55 @@
 # ref: https://en.wikipedia.org/wiki/AMD-V
 HW_CPU_X86_SVM = _CPU_X86_NS + 'svm'
 
+# All hypervisor specific features are prefixed with this namespace
+_VIRT_NS = 'virt:'
+
+# All libvirt specific features are prefixed with this namespace
+_VIRT_LIBVIRT_NS = _VIRT_NS + 'libvirt:'
+
+# All Hyper-V specific features are prefixed with this namespace
+_VIRT_HYPERV_NS = _VIRT_NS + 'hyperv:'
+
+_VIRT_HYPERV_VM_GENERATION = _VIRT_HYPERV_NS + 'vm_generation:'
+VIRT_HYPERV_VM_GENERATION_1 = -_VIRT_HYPERV_VM_GENERATION + '1'
+VIRT_HYPERV_VM_GENERATION_2 = -_VIRT_HYPERV_VM_GENERATION + '2'
+
+VIRT_HYPERV_SECURE_BOOT = _VIRT_HYPERV_NS + 'secure_boot'
+VIRT_HYPERV_SHIELDED_VMS =_VIRT_HYPERV_NS + 'shielded_vms'
+
+# RemoteFX suffix
+_VIRT_HYPERV_REMOTEFX = _VIRT_HYPERV_NS + 'remotefx:'
+
+# VRAM is requested in MB
+_VIRT_HYPERV_REMOTEFX_MEMORY = _VIRT_HYPERV_REMOTEFX + 'vram:'
+VIRT_HYPERV_REMOTEFX_MEMORY_64 = _VIRT_HYPERV_REMOTEFX_MEMORY + '64'
+VIRT_HYPERV_REMOTEFX_MEMORY_128 = _VIRT_HYPERV_REMOTEFX_MEMORY + '128'
+VIRT_HYPERV_REMOTEFX_MEMORY_256 = _VIRT_HYPERV_REMOTEFX_MEMORY + '256'
+VIRT_HYPERV_REMOTEFX_MEMORY_512 = _VIRT_HYPERV_REMOTEFX_MEMORY + '512'
+VIRT_HYPERV_REMOTEFX_MEMORY_1024 = _VIRT_HYPERV_REMOTEFX_MEMORY + '1024'
```

Comment:
What the placement API is used to determine is whether a provider has enough capacity to handle some requested amount of resource.

If a Hyper-V compute node has 2048MB of video ram available, it would have an inventory record for resource class VIDEO_MEMORY_MB with a total of 2048. An image or flavor might have a requested amount of remotefx vram of any of the amounts you list above but that won't affect the accounting of the resources.

The inventory record could be:

 resource_provider_id: <UUID of compute node>
 resource_class_id: VIDEO_MEMORY_MB
 total: 2048
 reserved: 0
 min_unit: 64
 max_unit: 1024
 step_size: 64

That would solve most of the problems you indicate here. Note that the step_size is used to evaluate whether the requested amount of that resource class falls within a particular "step". When evaluating if the resource provider can meet a requested amount, we filter out providers by doing the following:

 WHERE (total - reserved) * allocation_ratio >= $REQUESTED_AMOUNT
 AND min_unit >= $REQUESTED_AMOUNT
 AND max_unit <= $REQUESTED_AMOUNT
 AND $REQUESTED_AMOUNT % step_size == 0

In the example of your vram resource classes, setting the step_size to 64 would eliminate all inventory records when requesting "bad" amounts of VIDEO_MEMORY_MB such as 1023 or 2040. Unfortunately, a requested amount of, say, 768 would *not* eliminate any inventory records because 768 modulo 64 equals 0. 

To prevent this issue, we could also add an additional boolean field called "power_stepping" that would cause the requested amount to be evaluated against a *power* of the step_size field value instead of a *modulo* of the step_size.

For instance, let's say we had the following inventory record:


 resource_provider_id: <UUID of compute node>
 resource_class_id: VIDEO_MEMORY_MB
 total: 2048
 reserved: 0
 min_unit: 64
 max_unit: 1024
 step_size: 2
 power_stepping: true

Inventory records would then be eliminated using the following expression:


 WHERE (total - reserved) * allocation_ratio >= $REQUESTED_AMOUNT
 AND min_unit >= $REQUESTED_AMOUNT
 AND max_unit <= $REQUESTED_AMOUNT
 AND (
   (power_stepping == FALSE AND $REQUESTED_AMOUNT % step_size == 0)
   OR
   (power_stepping == TRUE AND (
      LOG($REQUESTED_AMOUNT) / LOG(step_size) == CAST(LOG($REQUESTED_AMOUNT) / LOG(step_size) AS INT)
      )
)

The above would filter out all inventories if the requested amount was not a power of 2.